ykllvmwrap/src/ykllvmwrap.cc

```diff
@@ -37,15 +52,107 @@ extern "C" char *__yk_llvmwrap_symbolizer_find_code_sym(
+extern "C" void *__ykllvmwrap_irtrace_compile(char *FuncNames[], size_t BBs[], size_t Len, char *SecPtr, size_t SecSize) {
+    LLVMContext Context;
+    auto DstMod = new Module("", Context);
+
+    // Set up new module for the trace.
+    llvm::FunctionType *FType = llvm::FunctionType::get(Type::getVoidTy(Context), false);
+	llvm::Function *DstFunc = llvm::Function::Create(FType, Function::InternalLinkage, TRACE_FUNC_NAME, DstMod);
+    DstFunc->setCallingConv(CallingConv::C);
+    auto DstBB = BasicBlock::Create(Context, "bb0", DstFunc);
+	llvm::IRBuilder<> Builder(Context);
+	Builder.SetInsertPoint(DstBB);
+
+    // Load the binary's bitcode.
+    auto SrcMod = load_module(Context, SecPtr, SecSize);
+
+    llvm::ValueToValueMapTy VMap;
+    bool Tracing = false;
+
+    // Iterate over the PT trace and stitch together all traced blocks.
     for (size_t Idx = 0; Idx < Len; Idx++) {
-        // FIXME populate.
-        //auto FuncName = FuncNames[Idx];
-        //auto BB = BBs[Idx];
+        auto FuncName = FuncNames[Idx];
+
+        // Get a traced function so we can extract blocks from it.
+        Function *F = SrcMod->getFunction(FuncName);
+        if (!F)
+            errx(EXIT_FAILURE, "can't find function %s", FuncName);
+
+        // Skip to the correct block.
+        auto It = F->begin();
+        std::advance(It, BBs[Idx]);
+        BasicBlock *BB = &*It;
+        // Iterate over all instructions within this block and copy them over
+        // to our new module.
+        for (auto I = BB->begin(), EI = BB->end(); I != EI; I++) {
+            if (isa<CallInst>(I)) {
+                Function *CF = cast<CallInst>(&*I)->getCalledFunction();
+                if (CF->getName() == "__yktrace_start_tracing") {
+                    Tracing = true;
+                    I++; // FIXME Better way of removing result of __yktrace_start_tracing.
+                    continue;
+                } else if (CF->getName() == "__yktrace_stop_tracing") {
+                    // FIXME Remove argument setup before __yktrace_stop_tracing call.
+                    Tracing = false;
+                }
+            }
+            if (!Tracing && !llvm::isa<AllocaInst>(I)) {
+                // FIXME Remove allocas not needed in traced code.
+                continue;
+            }
+            if (llvm::isa<llvm::BranchInst>(I)) {
+                // FIXME Replace all branch instruction with guards.
+                continue;
+            }
+
+            // Copy instruction over into the IR trace. Since the instruction
+            // operands still reference values in the original bitcode, remap
+            // the operands to point to new values within the IR trace.
+            auto NewInst = &*I->clone();
+            llvm::RemapInstruction(NewInst, VMap, RF_NoModuleLevelChanges);
+            VMap[&*I] = NewInst;
+            Builder.Insert(NewInst);
+        }
     }
+    Builder.CreateRetVoid();
+    // FIXME Only do this for debug builds.
```

Comment:
I think we should fix this in this PR (it's a one line fix).

Reply:
I don't think it is. We'd need to pass down the cargo profile to C++ somehow.

Reply:
It's fine to do something like `#ifdef DEBUG` or whatever we think we *will* be doing. I can be convinced otherwise, but this doesn't feel like it should be too difficult?

Reply:
I guess this is better than a comment, even if `DEBUG` isn't set anywhere yet: 208fcb6.

Reply:
Works for me.